ldurniat/berry.lua

```diff
@@ -447,484 +468,553 @@ end
+function Map:addLayer( name ) 
 
-		for i = 1, self.numChildren do
+	local layer = display.newGroup()
 
-			layer = self[i]
+	layer.name = name
 
-			for j = 1, layer.numChildren do
+	-- These are the defaults
+	layer.properties = {}
+	layer.type = 'objectgroup'
+	layer.alpha = 1
+	layer.isVisible = true
+	layer.offset_x = 0
+	layer.offset_y = 0
+	layer.objects = {}
 
-				object = layer[j]
+	self:insert( layer )
+	
+	return layer
+end  
 
-				if ( not name or object.name == name ) and ( not objType or object.type == objType ) then
+------------------------------------------------------------------------------------------------
+--- Find the layer by name.
+--
+-- @param name The name of layer.
+-- @return The layer object if found.
+-- 
+-- Original code from https://github.com/ponywolf/ponytiled 
+------------------------------------------------------------------------------------------------
+function Map:getLayer( name ) 
 
-					objects[#objects + 1] = object
+	local layer
 
-				end	
+	for i = 1, self.numChildren do
 
-			end	
+		layer = self[i]
 
-		end
+		if layer.name == name then
 
-		return objects
+			return layer
 
-	end  
+		end	
 
-	------------------------------------------------------------------------------------------------
-	--- Find the object by name and type.
-	--
-	-- @param options The table which contains two fields name and type.
-	-- @return The object if found.
-	-- 
-	-- Original code from https://github.com/ponywolf/ponytiled 
-	------------------------------------------------------------------------------------------------
-	function map:getObject( options ) 
+	end
+		
+end  
 
-		options = options or {}
+------------------------------------------------------------------------------------------------
+--- Find the objects by name and type.
+--
+-- @param options The table which contains two fields name and type.
+-- @return The table with found objects.
+-- 
+-- Original code from https://github.com/ponywolf/ponytiled 
+------------------------------------------------------------------------------------------------
+function Map:getObjects( options ) 
 
-		local name    = options.name
-		local objType = options.type
+	options = options or {}
 
-		local object, layer
+	local name    = options.name
+	local objType = options.type
 
-		for i = 1, self.numChildren do
+	local objects, object, layer = {}
 
-			layer = self[i]
+	for i = 1, self.numChildren do
 
-			for j = 1, layer.numChildren do
+		layer = self[i]
 
-				object = layer[j]
+		for j = 1, layer.numChildren do
 
-				if ( not name or object.name == name ) and ( not objType or object.type == objType ) then
+			object = layer[j]
 
-					return object
+			if name and objType then -- must match both
+				if ( object.name == name ) and ( object.type == objType ) then
 
-				end	
+					objects[#objects + 1] = object
 
-			end	
+				end
+			else  -- must match one
+				if ( object.name == name ) or ( object.type == objType ) then
```

Comment:
I would modify your condition like that 
````
if ( name and object.name == name ) or ( objType and object.type == objType ) then
````
since you can get objects you don't want e.g. tiles (they don't have `type` property in general).

Reply:
Ahhhh, nice catch.  Fixed with 53ede88